README.md
Improved Blizzard UI + (Battle for Azeroth Edition)
===================================================

This is a fork of the original [Improved Blizzard Ui](https://github.com/kaytotes/ImprovedBlizzardUI) by Kaytotes 
of which work has since ceased for the moment, and Improved Blizzard Ui Plus carries forward with improvements and continual bug fixes for future WoW Patch's

[Latest Releases](https://github.com/Misunderstood-Wookiee/ImprovedBlizzardUIPlus/releases/ "Download from GitHub Releases")

_________________________________________________________________________________________________________________
### Plus Edition? What gives.
As this suggests includes fixes for oversights & handles my personal gripes/desires. I do keep this updated as required via mergeing whenever Kytotes pushes fixes which do not directly conflict too maintain coherency between the original and this fork however due to Kaytots no longer providing support for BFA this version will likely be updated with better fixes, don't worry this version is still kept up-to date with the latest stability changes from Kytotos when they are made available.

The scope of this project/addon is staying true to the original where possible without major change to how the UI functions from the original, the original addon has been my primary UI for sometime now and I refuse to play without it.

*It appears Kytotes is no longer providing immediate support for BFA (shrug)*
I also pull request any major fixes over to Kaytoes's original project though when or if he accepts those and pushes a new build to curse/twitch is out of my control.
_________________________________________________________________________________________________________________
### Recommended Addons
Improved Blizzard UI + works best with the following addons.
* [Baud Bag](https://www.curseforge.com/wow/addons/baud-bag)
* [DejaCharacterStats](https://www.curseforge.com/wow/addons/dejacharacterstats)
* [Immersion](https://wow.curseforge.com/projects/immersion)
* [DynamicCam](https://wow.curseforge.com/projects/dynamiccam) + [My TrueCam Preset](https://pastebin.com/JY12CPND)
_________________________________________________________________________________________________________________
###### PLEASE DO NOT BUG KYTOTES WITH ANY ISSUE YOU FIND IN THIS FORK!!!


CHANGELOG.md
# Improved Blizzard UI +  Changelog

## 1.1.9
* Frame text size adjusted slightly for player, target, focus names.

## 1.1.8
* Fixed bug in ToT sometimes performing an invalid operation when getting class colours.

## 1.1.7
* Tooltip text sized increased to 13 from 12.
* Remnants of old addon name pre v1.1.5 corrected.
* Addon Config Menu command changed (/ibui+)
* Addon Config Readability Improved (spacing adjusted)
* New Config Options: Hide Background, Hide Arrows, Hide Stance Frame  (These can be found under Action Bars > Main Bar).
* Config Option "Action Bar Art" has been renamed and moved under the "Action Bars" Tab into its new home under Main Bar. (functionality is retained)
* Internal files have changed, endcaps.lua removed, mainbarart.lua added and framework has been improved to accomodate future main bar functionality.
* Change log clean-up.

## 1.1.6
* Fixed version Info showing wrong version.
* Guild Colour on Tooltips are now default something more vanilla-esk.

## 1.1.5
* Re-based Addon under new name formaly Improved Blizzard UI now Improved Blizzard UI +

## 1.1.4
* Framework added for Action Bar Art (Gryphon end-caps). It is now possible to remove the end caps from the action bar, more functions to come later!
* Target Frame, reduced font size to improve readability of large nameplates. 
* ToT Frame (Target of Target),  reduced font size to improve readability of large nameplates.

## 1.1.3 
* Minimap Zone Frame better fits the larger zone names.

## 1.1.2 
* Chat Channels now use uppercase characters ( for guild, party, e.t.c) & chat text box font size is slightly increased.

## 1.1.1 
* Re-applied adjustments from prev versions. (Brute Force Frame Pos??, I think not.)

## 1.1.0 
* Merged Kaytotes Improved Blizzard UI Master for 8.0 Maintence Fixes

## 1.0.5 
* Framework for Gryphon Art Hide added to config menu, Proper Frame Unlock and removed hacky fix.

## 1.0.1 
* Tweaks to frame & misc font default sizes for better fit, hacky unit frame unlock fix.

## 1.0.0 
* Unlinked interface font from chat. Re-posionted focus frame. 
* Project Start (BFA 8.0 Base)


modules/frames/focus.lua
--[[
    modules\frames\party.lua
    Styling and Repositioning of the Focus Frame
]]
local addonName, Loc = ...;

local FocusUnitFrame = CreateFrame('Frame', nil, UIParent);  

--[[
    Set the Buffs on top of the Frame
    @ return void
]]
local function SetBuffs()
    FocusFrame.buffsOnTop = true;
end

--[[
    Apply textures and fonts
    @ return void
]]
local function StyleFocusFrame()
    if (FramesDB.stylePrimaryFrames == false) then return; end

    if(UnitExists('focus') == false) then return; end

    local unitClassification = UnitClassification(FocusFrame.unit);

    -- Set Sizes
    FocusFrame.healthbar:SetHeight(29);
    FocusFrame.healthbar:SetPoint('TOPLEFT',7,-22);
    FocusFrame.healthbar.TextString:SetPoint('CENTER',-50,6);
    FocusFrame.deadText:SetPoint('CENTER',-50,6);
    FocusFrame.nameBackground:Hide();
    FocusFrame.Background:SetPoint('TOPLEFT',7,-22);

    -- Add Dragons etc if needed
    local frameTexture;
    if ( unitClassification == 'worldboss' or unitClassification == 'elite' ) then
		frameTexture = 'Interface\\Addons\\ImprovedBlizzardUIPlus\\media\\UI-TargetingFrame-Elite';
	elseif ( unitClassification == 'rareelite' ) then
		frameTexture = 'Interface\\Addons\\ImprovedBlizzardUIPlus\\media\\UI-TargetingFrame-Rare-Elite';
	elseif ( unitClassification == 'rare' ) then
        frameTexture = 'Interface\\Addons\\ImprovedBlizzardUIPlus\\media\\UI-TargetingFrame-Rare';
    else
        frameTexture = 'Interface\\Addons\\ImprovedBlizzardUIPlus\\media\\UI-TargetingFrame';
	end

    FocusFrame.borderTexture:SetTexture(frameTexture);

    if (FramesDB.focusClassColours) then
        Imp.ApplyClassColours(FocusFrame.healthbar, FocusFrame.healthbar.unit);
    end

    local file, size, flags = PlayerFrameHealthBarTextLeft:GetFont();
    local r, g, b, a = PlayerFrameHealthBarTextLeft:GetTextColor();

    FocusFrameTextureFrameName:SetFont(file, 10, flags);
    FocusFrameTextureFrameHealthBarText:SetTextColor(r, g, b, a);
    FocusFrameTextureFrameName:SetTextColor(r, g, b, a);

    FocusFrame.healthbar:SetWidth(119);
    FocusFrame.healthbar.lockColor = true;

    if (FocusFrameTextureFramePVPIcon:IsShown()) then
        FocusFrameTextureFramePVPIcon:Hide();
    end

    -- Style Font
    if(FocusFrameToT:IsShown()) then
        FocusFrameToTTextureFrameName:SetFont(ImpFont, 10, flags);
        FocusFrameToTTextureFrameName:SetTextColor(r, g, b, a);
    end

    SetBuffs();
end

--[[
    Moves the Focus Frame
    @ return void
]]
local function SetPosition()
    -- Position (Lets re-unlock the frame so it can be moved!)
    FocusFrame:SetMovable(true);
    FocusFrame:SetScale(FramesDB.primaryScale - 0.15);
end

--[[
    Don't allow rescaling
    @ return void
]]
local function FocusFrame_SetSmallSize_Hook(toggle)
    SetPosition();
    StyleFocusFrame();
end

--[[
    Handles the WoW API Events Registered Below

    @ param Frame $self The Frame that is handling the event 
    @ param string $event The WoW API Event that has been triggered
    @ param arg $... The arguments of the Event
    @ return void
]]
local function HandleEvents (self, event, ...)
    if (event == 'PLAYER_ENTERING_WORLD') then

        SetPosition();

        SetBuffs();

        StyleFocusFrame();
    end
end

-- Register the Modules Events
FocusUnitFrame:SetScript('OnEvent', HandleEvents);
FocusUnitFrame:RegisterEvent('PLAYER_ENTERING_WORLD');

hooksecurefunc('FocusFrame_UpdateBuffsOnTop', SetBuffs);
hooksecurefunc('FocusFrame_SetSmallSize', FocusFrame_SetSmallSize_Hook);

FocusFrame:HookScript('OnShow', function()
    StyleFocusFrame();
end);

FocusFrame:HookScript('OnUpdate', function()
    StyleFocusFrame();
end)

hooksecurefunc('UnitFrameHealthBar_Update', function(self)
    if (FramesDB.focusFrameClassColours and self.unit == 'focus') then
        Imp.ApplyClassColours(self, self.unit);
    end
end);
hooksecurefunc('HealthBar_OnValueChanged', function(self)
    if (FramesDB.focusFrameClassColours and self.unit == 'focus') then
        Imp.ApplyClassColours(self, self.unit);
    end
end);

modules/frames/target.lua
--[[
    modules\frames\player.lua
    Styles, Scales and Repositions the Player Unit Frame.
    Disables Player Portrait Spam
]]
local addonName, Loc = ...;

local TargetUnitFrame = CreateFrame('Frame', nil, UIParent);  

--[[
    Handles the actual styling and scaling of the Target frame.

    @ return void
]]
local function StyleTargetFrame()
    
    if (FramesDB.stylePrimaryFrames == false) then return; end

    local unitClassification = UnitClassification(TargetFrame.unit);

    -- Set Sizes
    TargetFrame.healthbar:SetHeight(29);
    TargetFrame.healthbar:SetPoint('TOPLEFT',7,-22);
    TargetFrame.healthbar.TextString:SetPoint('CENTER',-50,6);
    TargetFrame.deadText:SetPoint('CENTER',-50,6);
    TargetFrame.nameBackground:Hide();
    TargetFrame.Background:SetPoint('TOPLEFT',7,-22);

    -- Add Dragons etc if needed
    local frameTexture;
    if ( unitClassification == 'worldboss' or unitClassification == 'elite' ) then
		frameTexture = 'Interface\\Addons\\ImprovedBlizzardUIPlus\\media\\UI-TargetingFrame-Elite';
	elseif ( unitClassification == 'rareelite' ) then
		frameTexture = 'Interface\\Addons\\ImprovedBlizzardUIPlus\\media\\UI-TargetingFrame-Rare-Elite';
	elseif ( unitClassification == 'rare' ) then
        frameTexture = 'Interface\\Addons\\ImprovedBlizzardUIPlus\\media\\UI-TargetingFrame-Rare';
    else
        frameTexture = 'Interface\\Addons\\ImprovedBlizzardUIPlus\\media\\UI-TargetingFrame';
	end

    TargetFrame.borderTexture:SetTexture(frameTexture);

    if (FramesDB.targetClassColours) then
        Imp.ApplyClassColours(TargetFrame.healthbar, TargetFrame.healthbar.unit);
    end

    if (FramesDB.targetBuffsOnTop) then
        TargetFrame.buffsOnTop = true;
    end

    local file, size, flags = PlayerFrameHealthBarTextLeft:GetFont();
    local r, g, b, a = PlayerFrameHealthBarTextLeft:GetTextColor();

    TargetFrameTextureFrameHealthBarText:SetTextColor(r, g, b, a);
    TargetFrameTextureFrameName:SetTextColor(r, g, b, a);

    TargetFrameTextureFrameName:SetFont(ImpFont, 9, flags);

    TargetFrameTextureFrameHealthBarText:SetFont(ImpFont, 10, flags);
    TargetFrameTextureFrameHealthBarTextLeft:SetFont(ImpFont, 10, flags);
    TargetFrameTextureFrameHealthBarTextRight:SetFont(ImpFont, 10, flags);
    
    TargetFrameTextureFrameManaBarText:SetFont(ImpFont, 10, flags);
    TargetFrameTextureFrameManaBarTextLeft:SetFont(ImpFont, 10, flags);
    TargetFrameTextureFrameManaBarTextRight:SetFont(ImpFont, 10, flags);


    TargetFrame.healthbar:SetWidth(119);
    TargetFrame.healthbar.lockColor = true;

    if ( TargetFrame.totFrame ) then
        TargetFrameToTTextureFrameName:SetFont(file, 9, flags);
        TargetFrameToTTextureFrameName:SetTextColor(r, g, b, a);
    end
end

--[[
    Handles the WoW API Events Registered Below

    @ param Frame $self The Frame that is handling the event 
    @ param string $event The WoW API Event that has been triggered
    @ param arg $... The arguments of the Event
    @ return void
]]
local function HandleEvents (self, event, ...)
    if (event == 'PLAYER_ENTERING_WORLD') then
        -- Position (Lets re-unlock the frame so it can be moved!)
        TargetFrame:SetMovable(true);
        TargetFrame:SetScale(FramesDB.primaryScale);

        -- Style Frame
        StyleTargetFrame();
    end
end

-- Register the Modules Events
TargetUnitFrame:SetScript('OnEvent', HandleEvents);
TargetUnitFrame:RegisterEvent('PLAYER_ENTERING_WORLD');

function TargetofTargetHealthCheck_Hook(self)
    if (FramesDB.targetOfTargetClassColours) then
        Imp.ApplyClassColours(self.healthbar, self.healthbar.unit);
    end
end

-- Hook a bunch of Blizzard functions
hooksecurefunc('TargetFrame_CheckDead', StyleTargetFrame);
hooksecurefunc('TargetFrame_Update', StyleTargetFrame);
hooksecurefunc('TargetFrame_CheckFaction', StyleTargetFrame);
hooksecurefunc('TargetFrame_CheckClassification', StyleTargetFrame);
hooksecurefunc('TargetofTarget_Update', StyleTargetFrame);

hooksecurefunc('TargetofTargetHealthCheck', TargetofTargetHealthCheck_Hook);

hooksecurefunc('UnitFrameHealthBar_Update', function(self)
    if (FramesDB.targetOfTargetClassColours and self.unit == 'target') then
        Imp.ApplyClassColours(self, self.unit);
    end
end);
hooksecurefunc('HealthBar_OnValueChanged', function(self)
    if (FramesDB.targetOfTargetClassColours and self.unit == 'target') then
        Imp.ApplyClassColours(self, self.unit);
    end
end);

modules/frames/player.lua
--[[
    modules\frames\player.lua
    Styles, Scales and Repositions the Player Unit Frame.
    Disables Player Portrait Spam
]]
local addonName, Loc = ...;

local PlayerUnitFrame = CreateFrame('Frame', nil, UIParent);  

local confPlayerClassColours = false;

--[[
    Whenever the Portrait Damage text appears instantly override it

    @ param Frame $self The Frame that is handling the event
    @ param string $event The message eg 'IMMUNE', 'HEAL', 'BLOCK'
    @ param string $flags Extra information eg 'CRITICAL', 'RESIST'
    @ param int $amount THe amount of Damage or Healing done
    @ param int $type Honestly unsure, apparently sets it to black if > 0 but I've never seen this actually happen in game.
]]
local function CombatFeedback_OnCombatEvent_Hook(self, event, flags, amount, type)
    if(FramesDB.playerPortraitSpam) then
        self.feedbackText:SetText(' ');
    end
end

--[[
    Hides the Player Frame when you are out of combat, have no target and are at full health.

    @ param boolean $hide Should we hide the frame
    @ return void
]]
local function HidePlayer(hide)
    if (InCombatLockdown() == false and FramesDB.playerHideOOC) then
        if (hide and UnitHealth('player') == UnitHealthMax('player') and UnitExists('target') == false) then
            PlayerFrame:Hide();
        else
            PlayerFrame:Show();
        end
    end
end

local function StyleFrames()

    if (FramesDB.stylePrimaryFrames == false) then return end
    if (InCombatLockdown() == true) then return end

    PlayerFrameHealthBar:SetWidth(119);
    PlayerFrameHealthBar:SetHeight(29);
    PlayerFrameHealthBar:SetPoint('TOPLEFT',106,-22);
    PlayerFrameHealthBarText:SetPoint('CENTER',50,6);
    PlayerFrameTexture:SetTexture('Interface\\AddOns\\ImprovedBlizzardUIPlus\\media\\UI-TargetingFrame');
    PlayerStatusTexture:SetTexture('Interface\\AddOns\\ImprovedBlizzardUIPlus\\media\\UI-Player-Status');

    local file, size, flags = PlayerFrameHealthBarTextLeft:GetFont();
    local r, g, b, a = PlayerFrameHealthBarTextLeft:GetTextColor();

    PlayerName:SetFont(ImpFont, 10, flags);
    -- 10
    PlayerFrameHealthBarText:SetFont(ImpFont, 10, flags);
    PlayerFrameHealthBarTextLeft:SetFont(ImpFont, 10, flags);
    PlayerFrameHealthBarTextRight:SetFont(ImpFont, 10, flags);

    PlayerFrameManaBarText:SetFont(ImpFont, 10, flags);
    PlayerFrameManaBarTextLeft:SetFont(ImpFont, 10, flags);
    PlayerFrameManaBarTextRight:SetFont(ImpFont, 10, flags);

    PlayerFrameAlternateManaBarText:SetFont(ImpFont, 10, flags);
    PlayerFrameAlternateManaBar.RightText:SetFont(ImpFont, 10, flags);
    PlayerFrameAlternateManaBar.LeftText:SetFont(ImpFont, 10, flags);

    PlayerName:SetTextColor(r, g, b, a);

    PetName:SetFont(ImpFont, 11, flags);
    PetName:SetTextColor(r, g, b, a);
end

--[[
    Handles the WoW API Events Registered Below

    @ param Frame $self The Frame that is handling the event 
    @ param string $event The WoW API Event that has been triggered
    @ param arg $... The arguments of the Event
    @ return void
]]
local function HandleEvents (self, event, ...)
    if (event == 'PLAYER_LOGIN' or (event == 'ADDON_LOADED' and ... == 'ImprovedBlizzardUIPlus')) then
        confPlayerClassColours = FramesDB.playerClassColours;

        -- Position (Lets re-unlock the frame so it can be moved!)
        PlayerFrame:SetMovable(true);
        PlayerFrame:SetScale(FramesDB.primaryScale);

        HidePlayer(true);
        
        -- Style Frame
        if (FramesDB.stylePrimaryFrames) then
            StyleFrames();
        end
    end

    if (event == 'PLAYER_REGEN_DISABLED') then
        HidePlayer(false);
    elseif (event == 'PLAYER_REGEN_ENABLED' or event == 'UNIT_HEALTH') then
        HidePlayer(true);
    end

    if (event == 'PLAYER_TARGET_CHANGED') then
        if (UnitExists('target')) then
            HidePlayer(false);
        else
            HidePlayer(true);
        end
    end

    if (event == 'UNIT_EXITED_VEHICLE' and ... == 'player') then
        StyleFrames();
    end

    if (event == 'UNIT_ENTERED_VEHICLE' and ... == 'player') then
        StyleFrames();
    end
end

-- Register the Modules Events
PlayerUnitFrame:SetScript('OnEvent', HandleEvents);
PlayerUnitFrame:RegisterEvent('PLAYER_ENTERING_WORLD');
PlayerUnitFrame:RegisterEvent('PLAYER_LOGIN');
PlayerUnitFrame:RegisterEvent('UNIT_HEALTH');
PlayerUnitFrame:RegisterEvent('PLAYER_REGEN_DISABLED');
PlayerUnitFrame:RegisterEvent('PLAYER_REGEN_ENABLED');
PlayerUnitFrame:RegisterEvent('PLAYER_TARGET_CHANGED');
PlayerUnitFrame:RegisterEvent('UNIT_EXITED_VEHICLE');
PlayerUnitFrame:RegisterEvent('UNIT_ENTERED_VEHICLE');
PlayerUnitFrame:RegisterEvent('ADDON_LOADED');

-- Hook Blizzard Functions
hooksecurefunc('CombatFeedback_OnCombatEvent', CombatFeedback_OnCombatEvent_Hook);
hooksecurefunc('UnitFrameHealthBar_Update', function(self)
    if (confPlayerClassColours and self.unit == 'player') then
        Imp.ApplyClassColours(self, self.unit);
    end
end);
hooksecurefunc('HealthBar_OnValueChanged', function(self)
    if (confPlayerClassColours and self.unit == 'player') then
        Imp.ApplyClassColours(self, self.unit);
    end
end);

hooksecurefunc('PlayerFrame_Update', StyleFrames);